From 9bb15e8c871519813935099b4e132d390a9480cd Mon Sep 17 00:00:00 2001
From: Stephan Mueller <smueller@chronox.de>
Date: Tue, 27 Dec 2022 20:56:00 +0100
Subject: [PATCH 28/34] lrng: limit kernel crypto API default DRNGs

When the LRNG is compiled as the /dev/random provided (i.e. the legacy RNG
is not compiled by not selecting CONFIG_RANDOM_DEFAULT_IMPL), only the
ChaCha20 DRNG can be used as the default DRNG. This is due to the fact
that the LRNG must be capable of delivering random numbers very early
in the boot cycle. At that time, it is not guaranteed that the kernel
crypto API is available.

To select a kernel crypto API DRNG to be automatically loaded during boot,
select CONFIG_LRNG_SWITCH_DRBG=y. This selection allocates, for example,
the SP800-90A DRBG during boot, but at the time the kernel crypto API is
present.

To enforce this, CONFIG_LRNG_DFLT_DRNG_DRBG and CONFIG_LRNG_DFLT_DRNG_KCAPI
is only present if the legacy RNG is compiled and thus the LRNG does not
provide random numbers during early boot time.

Signed-off-by: Stephan Mueller <smueller@chronox.de>
---
 drivers/char/lrng/Kconfig | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/char/lrng/Kconfig b/drivers/char/lrng/Kconfig
index 08ab921a1..f4c343265 100644
--- a/drivers/char/lrng/Kconfig
+++ b/drivers/char/lrng/Kconfig
@@ -639,10 +639,12 @@ choice
 		select LRNG_DRNG_CHACHA20
 
 	config LRNG_DFLT_DRNG_DRBG
+		depends on RANDOM_DEFAULT_IMPL
 		bool "SP800-90A DRBG"
 		select LRNG_DRBG
 
 	config LRNG_DFLT_DRNG_KCAPI
+		depends on RANDOM_DEFAULT_IMPL
 		bool "Kernel Crypto API DRNG"
 		select LRNG_DRNG_KCAPI
 endchoice
-- 
2.39.0.rc2.1.gbd5df96b79

